src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -0,0 +1,99 @@
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Diagnostics.Tracing;
+using System.Reflection;
+
+namespace OpenTelemetry.Instrumentation.EventCounters;
+
+/// <summary>
+/// .NET EventCounters Instrumentation.
+/// </summary>
+internal class EventCountersMetrics : EventListener
+{
+    internal static readonly AssemblyName AssemblyName = typeof(EventCountersMetrics).Assembly.GetName();
+    internal static readonly Meter MeterInstance = new(AssemblyName.Name, AssemblyName.Version.ToString());
+
+    private readonly EventCountersInstrumentationOptions options;
+    private readonly ConcurrentBag<EventSource> preInitEventSources = new();
+    private readonly ConcurrentDictionary<Tuple<string, string>, Instrument> instruments = new();
+    private readonly ConcurrentDictionary<Tuple<string, string>, double> values = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="EventCountersMetrics"/> class.
+    /// </summary>
+    /// <param name="options">The options to define the metrics.</param>
+    public EventCountersMetrics(EventCountersInstrumentationOptions options)
+    {
+        this.options = options;
+
+        while (this.preInitEventSources.TryTake(out EventSource source))
+        {
+            if (this.options.ShouldListenToSource(source.Name))
+            {
+                this.EnableEvents(source, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventSourceCreated(EventSource source)
+    {
+        if (this.options == null)
+        {
+            this.preInitEventSources.Add(source);
+        }
+        else if (this.options.ShouldListenToSource(source.Name))
+        {
+            this.EnableEvents(source, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventWritten(EventWrittenEventArgs eventData)
+    {
+        if (this.options == null)
+        {
+            return;
+        }
+
+        var payload = eventData.Payload[0] as IDictionary<string, object>;
+        var name = payload["Name"].ToString();
+        var isGauge = payload.ContainsKey("Mean");
+        Tuple<string, string> metricKey = new(eventData.EventSource.Name, name);
+
+        if (isGauge)
+        {
+            this.values[metricKey] = Convert.ToDouble(payload["Mean"]);
+        }
+        else
+        {
+            _ = this.values.TryAdd(metricKey, 0);
+            this.values[metricKey] += Convert.ToDouble(payload["Increment"]);
+        }
+
+        _ = this.instruments.TryAdd(
+            metricKey,
+            isGauge ? MeterInstance.CreateObservableGauge(name, () => this.values[metricKey]) : MeterInstance.CreateObservableCounter(name, () => this.values[metricKey]));
```

Comment:
What if the `name` already exists under the `MeterInstance`?

Reply:
Currently it will not crash, but the value of the metric will be `0`.

I could add a warning that an instrument named that already exists?
Or probably better, is the name of the created instrument could be $"{eventSourceName}.${eventCounterName}"

Reply:
> I could add a warning that an instrument named that already exists? Or probably better, is the name of the created instrument could be "eventSourceName.{eventCounterName}"

The latter sounds good to me if we make sure `eventSourceName` is properly escaped (so we have a valid instrument name which follows the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-naming-rule)).

Doesn't have to block this PR though.

Reply:
Even if name doesn't already exist, if the EventSource name isn't going to be the Meter name it should probably be in the instrument name. Some counter names may be very ambiguous such as "size", "length", or "requests."

Reply:
Sounds good, yes I will include a warning and simply not add an additional instrument. The instrument for event source `a` with counter `b.c` will go to the `a.b.c` instrument, same as the event source `a.b` with counter `c`